### PR TITLE
Add initial jq-based templating engine

### PR DESCRIPTION
**Pull request by yosifkit**
Similar to https://github.com/docker-library/php/pull/1052.

This also takes some bits (like `dirCommit` and `Dockerfile.debian`) from https://github.com/docker-library/mysql/pull/680 in order to support that PR easier.

The new Dockerfile template should make it easier to keep the version-specific Dockerfiles in sync.

```diff
--- ../official-images/library/mysql	2020-10-22 15:16:01.132359192 -0700
+++ /dev/fd/63	2020-11-18 15:46:54.728546973 -0800
@@ -1,17 +1,20 @@
-# this file is generated via https://github.com/docker-library/mysql/blob/ff627942605ce08a32fa527993337df0e9856760/generate-stackbrew-library.sh
+# this file is generated via https://github.com/docker-library/mysql/blob/e0ac64268f0f2d151baea664b5fa9dbcaa0c9321/generate-stackbrew-library.sh
 
 Maintainers: Tianon Gravi <admwiggin@gmail.com> (@tianon),
              Joseph Ferguson <yosifkit@gmail.com> (@yosifkit)
 GitRepo: https://github.com/docker-library/mysql.git
 
 Tags: 8.0.22, 8.0, 8, latest
-GitCommit: 1a25a09159a3d7b6b5b8ad0e6c7eb53504b3aab5
+GitCommit: e0ac64268f0f2d151baea664b5fa9dbcaa0c9321
 Directory: 8.0
+File: Dockerfile.debian
 
 Tags: 5.7.32, 5.7, 5
-GitCommit: 1a25a09159a3d7b6b5b8ad0e6c7eb53504b3aab5
+GitCommit: e0ac64268f0f2d151baea664b5fa9dbcaa0c9321
 Directory: 5.7
+File: Dockerfile.debian
 
 Tags: 5.6.50, 5.6
-GitCommit: 1a25a09159a3d7b6b5b8ad0e6c7eb53504b3aab5
+GitCommit: e0ac64268f0f2d151baea664b5fa9dbcaa0c9321
 Directory: 5.6
+File: Dockerfile.debian
```

### CODE REVIEW

File: versions.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	export baseVersion debianSuite debianVersion
+	json="$(
+		jq <<<"$json" -c \
+			'.[env.version] = {
+				version: env.baseVersion,
+				debian: {
+					architectures: [ "amd64" ],
+					suite: env.debianSuite,
+					version: env.debianVersion,
+				}
```
Comment (tianon) [inline versions.sh:53]:
```suggestion
				},
```

:eyes:

**Review by tianon — COMMENTED** [review]
Overall looks good, nice work!

Just a few (minor) comments. :+1:

**Comment by tianon** [comment]
(I look forward to rebasing #680 :laughing:)

File: apply-templates.sh [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		# PLEASE DO NOT EDIT IT DIRECTLY.
+		#
+
+	EOH
+}
+
+for version; do
+	export version
+
+	if [ -n "$(jq -r '.[env.version].debian // ""' versions.json)" ]; then
```
[the remark this answers is not in the document]
Comment (yosifkit) [reply apply-templates.sh]: Yeah, the `if` was inspired by a similar section in 680: https://github.com/docker-library/mysql/pull/680/files#diff-e14594f6bb9b29ac0146c04e78ae33e8e92f2f00709fefab97359dea31b94d11R98-R127. It seems fine to keep.

**Comment by tianon** [comment]
Looks great -- actual `Dockerfile` changes are very minimal, and this improves the maintainability (especially the ability to verify PRs are updating templates properly and that the CI is testing the PR changes and that @docker-library-bot won't unexpectedly revert or change things post-merge).

**Pull request merged**